### ISSUE TITLE
[Bug] "(raylet) bash: line 0: exec: podman: not found" when using runtime_env container in cluster based on rayproject/ray image

### ISSUE THREAD

**Issue by efokschaner** [ray-project/ray#21617]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Clusters

### What happened + What you expected to happen

When I `ray submit` a job with a runtime_env using  `rayproject/ray:1.9.2-py39-cpu` container to a cluster based on the `rayproject/ray:1.9.2-py39-cpu` container, I get stuck on an error `(raylet) bash: line 0: exec: podman: not found`

Repro details included in other form field.

Output from `ray submit`
```
> ray submit cluster.yaml test.py 
2022-01-14 18:50:35,269 INFO util.py:282 -- setting max workers for head node type to 0
Loaded cached provider configuration
If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
Fetched IP: <HEAD IP>
Shared connection to <HEAD IP> closed.
Shared connection to <HEAD IP> closed.
2022-01-14 18:50:39,025 INFO util.py:282 -- setting max workers for head node type to 0
Fetched IP: <HEAD IP>
Shared connection to <HEAD IP> closed.
2022-01-14 18:50:42,641 INFO worker.py:842 -- Connecting to existing Ray cluster at address: <INTERNAL HEAD IP>:6379
(raylet) bash: line 0: exec: podman: not found
(raylet) bash: line 0: exec: podman: not found
(raylet) bash: line 0: exec: podman: not found
(raylet) bash: line 0: exec: podman: not found
(raylet) bash: line 0: exec: podman: not found
```

I would expect that ray core, having a dependency on podman, would have podman pre-installed within its own distribution image.

In my real use case, my `runtime_env` actually extends the `rayproject/ray:1.9.2-py39-cpu` container but just using the same container makes the repro clearer.

### Versions / Dependencies

I am using GCP with a `sourceImage: projects/cos-cloud/global/images/family/cos-stable`
I am running `rayproject/ray:1.9.2-py39-cpu` as shown in the below configs.

### Reproduction script

The following config + test script reproduces the issue for me:

In `cluster.yaml`:
```yaml
docker:
  image: "rayproject/ray:1.9.2-py39-cpu"
  container_name: "ray_container"
  # If true, pulls latest version of image. Otherwise, `docker run` will only pull the image
  # if no cached version is present.
  pull_before_run: True
  run_options:  # Extra options to pass into "docker run"
    - --ulimit nofile=65536:65536
```

In `test.py`:
```py
import ray


@ray.remote
def ray_main(echo: str):
    return echo


ray.init(
    address="auto",
    runtime_env={"container": {"image": "rayproject/ray:1.9.2-py39-cpu"}},
)

print(ray.get(ray_main.remote("Hello world")))
```

Then run:
```bash
ray submit cluster.yaml test.py
```

### Anything else

I'm going to try a workaround of extending `rayproject/ray:1.9.2-py39-cpu` to include podman.

Perhaps a fix for everyone might be to add `podman` to the packages here https://github.com/ray-project/ray/blob/cbcd03b77941e88a8b640b7b1c235de7847f79ed/docker/base-deps/Dockerfile#L30

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by efokschaner** [comment]
With podman installed in the ray container, i'm now seeing a different error:
```
2022-01-14 21:33:34,006 INFO worker.py:842 -- Connecting to existing Ray cluster at address: 10.182.0.7:6379
(raylet) time="2022-01-14T21:33:34-08:00" level=warning msg="\"/\" is not a shared mount, this could cause issues or missing mounts with rootless containers"
(raylet) cannot clone: Operation not permitted
(raylet) Error: cannot re-exec process
```

**Comment by efokschaner** [comment]
Based on https://www.redhat.com/sysadmin/podman-inside-container I tried adding the following `docker.run_options`:
```yaml
  run_options:  # Extra options to pass into "docker run"
    - --ulimit nofile=65536:65536
    - --cap-add SYS_ADMIN
    - --cap-add MKNOD
    - --security-opt seccomp=unconfined
    - --security-opt label=disable
```
I then run in to this error:
```
2022-01-15 00:38:01,401 INFO worker.py:842 -- Connecting to existing Ray cluster at address: <HEAD IP>:6379
(raylet) time="2022-01-15T00:38:02-08:00" level=warning msg="\"/\" is not a shared mount, this could cause issues or missing mounts with rootless containers"
(raylet) Error: mount /home/ray/.local/share/containers/storage/overlay:/home/ray/.local/share/containers/storage/overlay, flags: 0x1000: permission denied
```

I'm going to try removing the `runtime_env` option and instead use that image as my `docker.image` for the cluster.

**Comment by rkooo567** [comment]
cc @architkulkarni is it a known issue this doesn't work with podman?

**Comment by simon-mo** [comment]
Hi @efokschaner, thanks for the investigation. The container integration is still experimental. As you mentioned, currently we have a hard dependency on podman. I had success with just `docker run --privileged` to run the outer container.

**Comment by efokschaner** [comment]
Thanks @simon-mo 

I may give that a try, would --privileged be the recommended way to run the cluster's runtime in the future?

And do you have plans to include Podman within your rayproject/ray container in the future?

**Comment by simon-mo** [comment]
We will definitely have answer to both when we graduate the feature from experimental to stable. For your use case, do you have a constraint on the privileged flag?

**Comment by efokschaner** [comment]
Thanks. My use case is just experimenting with the framework and so I have unconstrained access to root on the cluster.

**Comment by vicyap** [comment]
Hi, I just ran into this issue. What does this mean?
> I had success with just docker run --privileged to run the outer container.

Is it possible to change the docker run options if I've deployed Ray on Kubernetes?

**Comment by simon-mo** [comment]
For K8s you will need to add SYS_ADMIN to capabilities field in the pod spec.

**Comment by ecm200** [comment]
Installing of `podman` into the container image, and running with the `--priveleged` option on the outer docker container has got me so far. But I have run up against a problem.

Does anyone have a feel for how `podman` works with the need to authenticate to private container repositories?

I have managed to get the `ray` cluster on Azure VMs to pull container images from a privately hosted image repository during the node startup by using the Azure CLI tools `az login` and `az acr login` in the `initialization_commands` of the `cluster yaml config`. This authenticates to the Azure system using a [user assigned managed identity ](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-authentication-managed-identity#example-1-access-with-a-user-assigned-identity). As I say, I have got this working by executing these commands **BEFORE** the images are pulled.

When specifying the images at execution time using the `ray.init(runtime_env=)` argument using a docker container, I get an authentication error.

Any ideas how this could be achieved?
Is it possible to run init commands prior to the pulling of the images?
